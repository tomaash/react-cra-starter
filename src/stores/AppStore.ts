import { observable } from 'mobx';
import ky from 'ky'
import { History, HistoryListenerParameter } from '@reach/router';

interface User {
  id: number
  first_name: string
  last_name: string
  avatar: string
}

interface ApiResponse {
  page: 1
  per_page: 3
  total: 12
  total_pages: 4
}

interface UserResponse extends ApiResponse {
  data: User[]
}

export class AppStore {

  @observable users: User[] = [];
  @observable currentRoute: HistoryListenerParameter

  setupHistory(history: History) {
    console.log(history)
    history.listen((state)=>{
      console.log(state)
      this.currentRoute = state
    })
  }

  loadUsers = async () => {
    const res: UserResponse = await ky.get('https://reqres.in/api/users').json()
    this.users = res.data
  }

}
