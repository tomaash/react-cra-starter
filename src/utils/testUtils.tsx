import React from 'react'
import ReactDOMServer from 'react-dom/server'
const path = require('path')

export const squashHtmlTags = str => str.replace(/(<[^>]+>)+/g, '\n')

export const renderText = element =>
  squashHtmlTags(ReactDOMServer.renderToStaticMarkup(element))

export const startPolly = dirname => {
  const { Polly } = require('@pollyjs/core')
  const { setupPolly } = require('setup-polly-jest')
  const FSPersister = require('@pollyjs/persister-fs')
  const XHRAdapter = require('@pollyjs/adapter-xhr')

  Polly.register(XHRAdapter)
  Polly.register(FSPersister)

  setupPolly({
    adapters: ['xhr'],
    recordIfMissing: true,
    persister: 'fs',
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(dirname, '__recordings__')
      }
    }
  })
}

export const sleep = m => new Promise(r => setTimeout(r, m))
