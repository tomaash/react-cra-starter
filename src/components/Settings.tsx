import React, { Component } from 'react'
import { observer, inject } from 'mobx-react'
import validatorjs from 'validatorjs'
import MobxReactForm from 'mobx-react-form'
import dvr from 'mobx-react-form/lib/validators/DVR'
import { MaterialTextField } from './MaterialTextField'
import { RouteComponentProps } from '@reach/router'
import { Button, Snackbar, IconButton } from '@material-ui/core'
import { AppStore } from '~/stores/AppStore'
import { observable } from 'mobx'
import { SnackbarPopup } from './shared/SnackbarPopup'

interface SettingsProps extends RouteComponentProps {
  appStore?: AppStore
}

@inject('appStore')
@observer
export class Settings extends Component<SettingsProps> {
  @observable popupConfig = {
    message: 'Settings updated',
    open: false
  }

  formHooks = {
    onError: form => {
      console.error('Form Error')
    },
    onSuccess: async form => {
      const values = form.values()
      await this.props.appStore.saveSettings(values)
      this.popupConfig.open = true
    }
  }

  formFields = {
    first_name: {
      label: 'First Name',
      placeholder: 'Insert First Name',
      rules: 'required|string|between:5,25',
      default: 'John'
    },
    last_name: {
      label: 'Last Name',
      placeholder: 'Insert Last Name',
      rules: 'required|string|between:5,25',
      default: 'Doe'
    }
  }
  form = new MobxReactForm(
    { fields: this.formFields },
    { plugins: { dvr: dvr(validatorjs) }, hooks: this.formHooks }
  )
  render() {
    return (
      <div className='p-8'>
        <form onSubmit={this.form.onSubmit}>
          <MaterialTextField
            testId='first_name'
            field={this.form.$('first_name')}
          />
          <MaterialTextField
            testId='last_name'
            field={this.form.$('last_name')}
          />

          <br />
          <Button
            data-testid='submit_button'
            className='mr-2'
            variant='outlined'
            color='primary'
            onClick={this.form.onSubmit}
          >
            Submit
          </Button>
          <Button
            className='mr-2'
            variant='outlined'
            color='secondary'
            onClick={this.form.onClear}
          >
            Clear
          </Button>
          <Button variant='outlined' onClick={this.form.onReset}>
            Reset
          </Button>

          <p>{this.form.error}</p>
        </form>
        <SnackbarPopup config={this.popupConfig} />
      </div>
    )
  }
}
