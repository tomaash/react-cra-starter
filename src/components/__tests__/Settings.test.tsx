import React from 'react'
import { shallow, mount } from 'enzyme'
import { renderText, sleep } from '~/utils/testUtils'
import { Settings } from '../Settings'
import { AppStore } from '~/stores/AppStore'

describe('<Settings />', () => {
  it('renders component', () => {
    const text = renderText(<Settings appStore={{} as any} />)
    expect(text).toMatchSnapshot()
  })

  it('submits valid form', async () => {
    const appStore = new AppStore()
    appStore.saveSettings = jest.fn()

    const wrapper = mount(<Settings appStore={appStore} />)

    const fnameInput = wrapper.find('[data-testid="first_name"]')
    fnameInput.simulate('change', { target: { value: 'MyFirstName' } })

    const lnameInput = wrapper.find('[data-testid="last_name"]')
    lnameInput.simulate('change', { target: { value: 'MyLastName' } })

    const sButton = wrapper.find('[data-testid="submit_button"]').first()
    sButton.simulate('click')
    await sleep(0)
    expect(appStore.saveSettings).toHaveBeenCalledWith({
      first_name: 'MyFirstName',
      last_name: 'MyLastName'
    })
  })

  it('does not submit invalid form', async () => {
    const appStore = new AppStore()
    appStore.saveSettings = jest.fn()

    const wrapper = mount(<Settings appStore={appStore} />)

    const fnameInput = wrapper.find('[data-testid="first_name"]')
    fnameInput.simulate('change', { target: { value: 'foo' } })

    const lnameInput = wrapper.find('[data-testid="last_name"]')
    lnameInput.simulate('change', { target: { value: 'bar' } })

    const sButton = wrapper.find('[data-testid="submit_button"]').first()
    sButton.simulate('click')
    await sleep(0)
    expect(appStore.saveSettings).not.toHaveBeenCalled()
  })
})
