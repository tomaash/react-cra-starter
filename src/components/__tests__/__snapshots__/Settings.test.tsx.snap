// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> renders component 1`] = `
"
First Name
Last Name
Submit
Clear
Reset
"
`;
